pkg/protocol/tipmanager/v1/tipmanager.go

```diff
@@ -0,0 +1,363 @@
+package tipmanagerv1
+
+import (
+	"sync"
+
+	"golang.org/x/xerrors"
+
+	"github.com/iotaledger/hive.go/ds/advancedset"
+	"github.com/iotaledger/hive.go/ds/randommap"
+	"github.com/iotaledger/hive.go/ds/shrinkingmap"
+	"github.com/iotaledger/hive.go/lo"
+	"github.com/iotaledger/hive.go/runtime/event"
+	"github.com/iotaledger/hive.go/runtime/module"
+	"github.com/iotaledger/hive.go/runtime/options"
+	"github.com/iotaledger/iota-core/pkg/model"
+	"github.com/iotaledger/iota-core/pkg/protocol/engine"
+	"github.com/iotaledger/iota-core/pkg/protocol/engine/blocks"
+	"github.com/iotaledger/iota-core/pkg/protocol/engine/booker"
+	"github.com/iotaledger/iota-core/pkg/protocol/engine/mempool"
+	"github.com/iotaledger/iota-core/pkg/protocol/engine/mempool/conflictdag"
+	"github.com/iotaledger/iota-core/pkg/protocol/tipmanager"
+	iotago "github.com/iotaledger/iota.go/v4"
+)
+
+// TipManager is a component that manages the tips of the Tangle.
+type TipManager struct {
+	// retrieveBlock is a function that retrieves a Block from the Tangle.
+	retrieveBlock func(blockID iotago.BlockID) (block *blocks.Block, exists bool)
+
+	// retrieveRootBlocks is a function that returns the current root blocks.
+	retrieveRootBlocks func() iotago.BlockIDs
+
+	// conflictDAG is the ConflictDAG that is used to track conflicts.
+	conflictDAG conflictdag.ConflictDAG[iotago.TransactionID, iotago.OutputID, booker.BlockVotePower]
+
+	// memPool holds information about pending transactions.
+	memPool mempool.MemPool[booker.BlockVotePower]
+
+	// tipMetadataStorage contains the TipMetadata of all Blocks that are managed by the TipManager.
+	tipMetadataStorage *shrinkingmap.ShrinkingMap[iotago.SlotIndex, *shrinkingmap.ShrinkingMap[iotago.BlockID, *TipMetadata]]
```

Comment:
Because it does not expose any of the thread safe methods of the ShrinkingMap (i.e. DeleteAndReturn, and so on and the only thing that the IndexedStorage will bring to the table would be a different interface for eviction, which is currently not horribly complicated either:

```
if evictedObjects, deleted := t.tipMetadataStorage.DeleteAndReturn(slotIndex); deleted {
	evictedObjects.ForEach(func(_ iotago.BlockID, tipMetadata *TipMetadata) bool {
		// ...
	})
}
```